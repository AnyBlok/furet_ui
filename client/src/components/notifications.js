/*
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
*/

import { defineComponent } from './factory';

defineComponent('notifications', {
  template: `
    <div v-bind:style="{position: 'fixed', right: '10px', top: '60px', width: '400px', zIndex: 1000}">
      <component 
        v-for="notification in notifications"
        is="notification"
        v-bind:key="notification.id"
        v-bind:id="notification.id"
        v-bind:type="notification.notification_type"
        v-bind:title="notification.title"
        v-bind:message="notification.message"
        v-bind:has_icon="notification.has_icon"
      />
    </div>`,
  prototype: {
    computed: {
      notifications() {
        return this.$store.state.global.notifications;
      },
    },
  },
});

defineComponent('notification', {
  template: `
    <b-message 
      v-bind:title="title" 
      v-bind:has-icon="has_icon" 
      v-bind:type="getType"
      v-on:close="onClose"
    >
      <div v-html="message" />
    </b-message>`,
  prototype: {
    props: ['id', 'type', 'title', 'message', 'has_icon'],
    computed: {
      getType() {
        switch (this.type) {
          case 'success':
            return 'is-success';
          case 'info':
            return 'is-info';
          case 'warning':
            return 'is-warning';
          case 'error':
            return 'is-danger';
          default:
            return null;
        }
      },
    },
    methods: {
      onClose() {
        this.$store.commit('REMOVE_NOTIFICATION', { id: this.id });
      },
    },
  },
});

