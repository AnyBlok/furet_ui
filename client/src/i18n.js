import Vue from 'vue';
import VueI18n from 'vue-i18n';

Vue.use(VueI18n);

export const i18nConf = {
  locale: 'en',
  fallbackLocale: 'en',
  messages: {
    en: {
    },
  },
};
export const i18n = new VueI18n(i18nConf);
