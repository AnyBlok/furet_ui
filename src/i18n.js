/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import VueI18n from 'vue-i18n';
Vue.use(VueI18n);


export const i18nConf = {
    locale: 'en',
    fallbackLocale: 'en',
    messages: {
        en: {
            menus: {
                close: 'Close',
                search: 'Search ...',
            },
            views: {
                unknown: {
                    title: 'The wanted view "{name}" is unknown',
                    message: "Report this message to the administrator",
                },
                common: {
                    create: 'Create',
                    save: 'Save',
                    edit: 'Edit',
                    cancel: 'Cancel',
                    delete: 'Delete',
                    close: 'Close',
                    actions: 'Actions',
                    more: 'More',
                },
                clients: {
                    login: {
                        button: 'Log in',
                    },
                },
            },
            fields: {
                common: {
                    required: 'this field is required',
                },
                json: {
                    invalid: 'Invalid JSON format',
                },
                file: {
                    upload: 'Upload',
                    download: 'Download',
                    delete: 'Delete',
                },
            },
        },
    },
};
export const i18n = new VueI18n(i18nConf);
