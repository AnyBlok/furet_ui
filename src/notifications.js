/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';

export const Notifications = Vue.component('notifications', {
    template: `
        <div
            v-bind:style="{position: 'fixed', right: '10px', top: '60px', width: '400px', zIndex: 1000}"
        >
            <component 
                v-for="notification in notifications"
                is="notification"
                v-bind:key="notification.id"
                v-bind:id="notification.id"
                v-bind:type="notification.notification_type"
                v-bind:title="notification.title"
                v-bind:message="notification.message"
                v-bind:has_icon="notification.has_icon"
            />
        </div>`,
    computed: {
        notifications () {
            return this.$store.state.global.notifications;
        },
    },
})

export const Notification = Vue.component('notification', {
    props: ['id', 'type', 'title', 'message', 'has_icon'],
    template: `
        <b-message 
            v-bind:title="title" 
            v-bind:has-icon="has_icon" 
            v-bind:type="getType"
            v-on:close="onClose"
        >
            <div v-html="message" />
        </b-message>`,
    computed: {
        getType () {
            switch (this.type) {
                case 'success':
                    return 'is-success';
                case 'info':
                    return 'is-info';
                case 'warning':
                    return 'is-warning';
                case 'error':
                    return 'is-danger';
                default:
                    return null;
            }
        }
    },
    methods: {
        onClose () {
            this.$store.commit('REMOVE_NOTIFICATION', {id: this.id});
        }
    }
})
