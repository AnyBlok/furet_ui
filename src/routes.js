/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import VueRouter from 'vue-router';
Vue.use(VueRouter);
import {json_post_dispatch_all} from './server-call';


const CustomeView = {
    template: '<furet-ui-custom-view v-bind:viewName="viewName" />',
    props: ['viewName'],
    beforeRouteEnter: (to, from, next) => {
        json_post_dispatch_all('/custom/view/' + to.params.viewName, {});
        next();
    },
    beforeRouteUpdate: (to, from, next) => {
        if (to.params.viewName != from.params.viewName) {
            json_post_dispatch_all('/custom/view/' + to.params.viewName, {});
        }
        next();
    },
}

const Action = {
    template: '<router-view></router-view>',
    beforeRouteEnter: (to, from, next) => {
        json_post_dispatch_all('/action/' + to.params.actionId, to.params);
        next();
    },
    beforeRouteUpdate: (to, from, next) => {
        if (to.params.actionId != from.params.actionId)
            json_post_dispatch_all('/action/' + to.params.actionId, to.params);
        next();
    },
};

const View = {
    template: `<furet-ui-view
                    v-bind:spaceId="spaceId"
                    v-bind:menuId="menuId"
                    v-bind:actionId="actionId"
                    v-bind:viewId="viewId"
                    v-bind:dataId="dataId"
                    v-bind:mode="mode"
                />`,
    props: ['spaceId', 'menuId', 'actionId', 'viewId', 'dataId', 'mode'],
    beforeRouteEnter: (to, from, next) => {
        json_post_dispatch_all('/view/' + to.params.viewId, to.params);
        next();
    },
    beforeRouteUpdate: (to, from, next) => {
        if (to.params.viewId != from.params.viewId)
            json_post_dispatch_all('/view/' + to.params.viewId, to.params);
        next();
    },
};


export const router = new VueRouter({
    routes: [
        {
            path: '/custom/view/:viewName',
            name: 'custom_view',
            props: true,
            component: CustomeView,
        },
        {
            path: '/space/:spaceId',
            name: 'space',
            props: true,
            component: {
                template: `<furet-ui-space
                                v-bind:spaceId="spaceId"
                                v-bind:menuId="menuId"
                                v-bind:actionId="actionId"
                                v-bind:viewId="viewId"
                            />`,
                props: ['spaceId', 'menuId', 'actionId', 'viewId'],
                beforeRouteEnter: (to, from, next) => {
                    json_post_dispatch_all('/space/' + to.params.spaceId, to.params);
                    next();
                },
                beforeRouteUpdate: (to, from, next) => {
                    if (to.params.spaceId != from.params.spaceId) json_post_dispatch_all('/space/' + to.params.spaceId, to.params);
                    next();
                },
            },
            children: [
                {
                    path: 'menu/:menuId',
                    name: 'space_menu',
                    component: {
                        template: '<router-view></router-view>'
                    },
                    children: [
                        {
                            path: 'action/:actionId',
                            name: 'space_menu_action',
                            component: Action,
                            children: [
                                {
                                    path: 'view/:viewId',
                                    name: 'space_menu_action_view',
                                    props: true,
                                    component: View,
                                    children: [
                                        {
                                            path: 'data/:dataId/mode/:mode',
                                            name: 'space_menu_action_view_dataId',
                                            component: {template: '<div />'},  // never called
                                        },
                                    ],
                                },
                                {
                                    path: 'custom/view/:viewName',
                                    name: 'space_menu_action_custom_view',
                                    props: true,
                                    component: CustomeView,
                                },
                            ],
                        },
                    ],
                },
                {
                    path: 'action/:actionId',
                    name: 'space_action',
                    component: Action,
                    children: [
                        {
                            path: 'view/:viewId',
                            name: 'space_action_view',
                            props: true,
                            component: View,
                            children: [
                                {
                                    path: 'data/:dataId/mode/:mode',
                                    name: 'space_action_view_dataId',
                                    component: {template: '<div />'},  // never called
                                },
                            ],
                        },
                        {
                            path: 'custom/view/:viewName',
                            name: 'space_action_custom_view',
                            props: true,
                            component: CustomeView,
                        },
                    ],
                },
            ]
        },
    ],
});

export default router;
