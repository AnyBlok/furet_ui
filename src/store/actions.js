/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
export const UNITEST_CLEAR = ({ commit }) => {
    commit('CLEAR_CLIENT');
    commit('CLEAR_DATA');
    commit('CLEAR_GLOBAL');
    commit('CLEAR_LEFT_MENU');
    commit('CLEAR_RIGHT_MENU');
}
