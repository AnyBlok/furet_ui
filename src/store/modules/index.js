/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import global from './global';
import client from './client';
import rightmenu from './right-menu';
import leftmenu from './left-menu';
import data from './data';

export default {
    global,
    client,
    rightmenu,
    leftmenu,
    data,
}
