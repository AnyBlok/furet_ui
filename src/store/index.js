/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import _ from 'underscore';
import Vue from 'vue';
import Vuex from 'vuex';
import * as actions from './actions';
import * as getters from './getters';
import modules from './modules';
import {i18n} from '../i18n';
import {router} from '../routes';
import moment from 'moment';

Vue.use(Vuex)
const debug = process.env.NODE_ENV !== 'production'

export const storeDef = {
    actions,
    getters,
    modules,
    strict: debug,
};

export const store = new Vuex.Store(storeDef);

export const dispatchAll = (datas) => {
    _.each(datas, data => {
        if (data.type) {
            switch (data.type) {
                case 'UPDATE_LOCALES':
                    _.each(data.locales, locale => {
                        i18n.setLocaleMessage(locale.locale, locale.messages)
                    });
                    break
                case 'SET_LOCALE':
                    i18n.locale = data.locale;
                    moment.locale(data.locale);
                    break
                case 'UPDATE_ROUTE':
                    if (data.name) router.push({name: data.name, params: data.params});
                    else if (data.path) router.push({path: data.path});
                    break;
                default:
                    const type = data.type;
                    delete data.type;
                    const process = data.process || 'commit';
                    store[process](type, data);
            }
        }
    });
}

export default store;
