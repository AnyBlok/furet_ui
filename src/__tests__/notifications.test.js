/**
This file is a part of the FuretUI project

   Copyright (C) 2017 Jean-Sebastien SUZANNE <jssuzanne@anybox.fr>

This Source Code Form is subject to the terms of the Mozilla Public License,
v. 2.0. If a copy of the MPL was not distributed with this file,You can
obtain one at http://mozilla.org/MPL/2.0/.
**/
import Vue from 'vue';
import {store} from '../store';
import {router} from '../routes';
import {i18n} from '../i18n';
import Buefy from 'buefy';
Vue.use(Buefy, {defaultIconPack: 'fa',});
import {Notifications, Notification} from '../notifications';

describe('Notifications component', () => {
    const renderer = require('vue-server-renderer').createRenderer();
    beforeEach(() => {
        store.dispatch('UNITEST_CLEAR');
    });
    it('notifications', () => {
        const vm = new Vue({
            el: document.createElement('div'),
            store,
            router,
            i18n,
            render: h => h(Notifications),
        });
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
        store.dispatch('ADD_NOTIFICATION', {title: 'Title 1', message: 'Message 1', process: 'dispatch'});
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
        store.dispatch('ADD_NOTIFICATION', {title: 'Title 2', message: 'Message 2', process: 'dispatch'});
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
    });
    it('notification', () => {
        const vm = new Vue({
            el: document.createElement('div'),
            store,
            router,
            i18n,
            render: h => h(Notification, {props: {
                title: 'Title 1',
                message: '<span>Message 1</span>',
            }}),
        });
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
    });
    it('notification success with icon', () => {
        const vm = new Vue({
            el: document.createElement('div'),
            store,
            router,
            i18n,
            render: h => h(Notification, {props: {
                title: 'Title 1',
                message: '<span>Message 1</span>',
                has_icon: true,
                type: 'success',
            }}),
        });
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
    });
    it('notification info with icon', () => {
        const vm = new Vue({
            el: document.createElement('div'),
            store,
            router,
            i18n,
            render: h => h(Notification, {props: {
                title: 'Title 1',
                message: '<span>Message 1</span>',
                has_icon: true,
                type: 'info',
            }}),
        });
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
    });
    it('notification warning with icon', () => {
        const vm = new Vue({
            el: document.createElement('div'),
            store,
            router,
            i18n,
            render: h => h(Notification, {props: {
                title: 'Title 1',
                message: '<span>Message 1</span>',
                has_icon: true,
                type: 'warning',
            }}),
        });
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
    });
    it('notification error with icon', () => {
        const vm = new Vue({
            el: document.createElement('div'),
            store,
            router,
            i18n,
            render: h => h(Notification, {props: {
                title: 'Title 1',
                message: '<span>Message 1</span>',
                has_icon: true,
                type: 'error',
            }}),
        });
        renderer.renderToString(vm, (err, str) => {
            expect(str).toMatchSnapshot();
        });
    });
});
