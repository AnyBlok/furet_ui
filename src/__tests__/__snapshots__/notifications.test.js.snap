// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications component notification 1`] = `
<article data-server-rendered="true" class="message">
    <header class="message-header">
        <p>Title 1</p> <button type="button" class="delete"></button></header>
    <section class="message-body">
        <div class="media">
            <!---->
            <div class="media-content">
                <div><span>Message 1</span></div>
            </div>
        </div>
    </section>
</article>
`;

exports[`Notifications component notification error with icon 1`] = `
<article data-server-rendered="true" class="message is-danger">
    <header class="message-header">
        <p>Title 1</p> <button type="button" class="delete"></button></header>
    <section class="message-body">
        <div class="media">
            <div class="media-left"><span class="icon is-large is-danger"><i class="fa fa-exclamation-circle"></i></span></div>
            <div class="media-content">
                <div><span>Message 1</span></div>
            </div>
        </div>
    </section>
</article>
`;

exports[`Notifications component notification info with icon 1`] = `
<article data-server-rendered="true" class="message is-info">
    <header class="message-header">
        <p>Title 1</p> <button type="button" class="delete"></button></header>
    <section class="message-body">
        <div class="media">
            <div class="media-left"><span class="icon is-large is-info"><i class="fa fa-info-circle"></i></span></div>
            <div class="media-content">
                <div><span>Message 1</span></div>
            </div>
        </div>
    </section>
</article>
`;

exports[`Notifications component notification success with icon 1`] = `
<article data-server-rendered="true" class="message is-success">
    <header class="message-header">
        <p>Title 1</p> <button type="button" class="delete"></button></header>
    <section class="message-body">
        <div class="media">
            <div class="media-left"><span class="icon is-large is-success"><i class="fa fa-check-circle"></i></span></div>
            <div class="media-content">
                <div><span>Message 1</span></div>
            </div>
        </div>
    </section>
</article>
`;

exports[`Notifications component notification warning with icon 1`] = `
<article data-server-rendered="true" class="message is-warning">
    <header class="message-header">
        <p>Title 1</p> <button type="button" class="delete"></button></header>
    <section class="message-body">
        <div class="media">
            <div class="media-left"><span class="icon is-large is-warning"><i class="fa fa-exclamation-triangle"></i></span></div>
            <div class="media-content">
                <div><span>Message 1</span></div>
            </div>
        </div>
    </section>
</article>
`;

exports[`Notifications component notifications 1`] = `<div data-server-rendered="true" style="position:fixed;right:10px;top:60px;width:400px;z-index:1000;"></div>`;

exports[`Notifications component notifications 2`] = `
<div data-server-rendered="true" style="position:fixed;right:10px;top:60px;width:400px;z-index:1000;">
    <article class="message">
        <header class="message-header">
            <p>Title 1</p> <button type="button" class="delete"></button></header>
        <section class="message-body">
            <div class="media">
                <!---->
                <div class="media-content">
                    <div>Message 1</div>
                </div>
            </div>
        </section>
    </article>
</div>
`;

exports[`Notifications component notifications 3`] = `
<div data-server-rendered="true" style="position:fixed;right:10px;top:60px;width:400px;z-index:1000;">
    <article class="message">
        <header class="message-header">
            <p>Title 1</p> <button type="button" class="delete"></button></header>
        <section class="message-body">
            <div class="media">
                <!---->
                <div class="media-content">
                    <div>Message 1</div>
                </div>
            </div>
        </section>
    </article>
    <article class="message">
        <header class="message-header">
            <p>Title 2</p> <button type="button" class="delete"></button></header>
        <section class="message-body">
            <div class="media">
                <!---->
                <div class="media-content">
                    <div>Message 2</div>
                </div>
            </div>
        </section>
    </article>
</div>
`;
